treeppl/tree-inference-naive.tppl

```diff
@@ -0,0 +1,311 @@
+node = function(left: Tree | Leaf, right: Tree | Leaf, age: NonnegativeReal, states: State[]) => Tree
+{
+    return {
+	type: Node,
+	left: left,
+	right: right,
+	age: age,
+	states: states
+    }
+}
+
+/** Convert tree to Newick */
+top_to_newick = function(t: Tree)
+{
+    // TODO not implemented
+}
+
+/// Core model functions ///
+
+
+/** Recursive function to randomly build a tree */
+cluster = function(
+    Q: InstantaneousRateMatrix,
+    trees: Array<Tree>,
+    max_age : Age
+) => Array<Tree>
+{
+    // Check if we are done
+    if trees.length == 1
+        return trees
+
+    // Pick two random trees in the trees vector
+    v ~ Permutation(0:(trees.length - 1))
+    i = v[0]
+    j = v[1]
+
+    // Propose s (new node states) from arbitrary proposal distribution (here stationary distribution)
+    propose s ~ iid(
+	dist = Categorical(probabilities = stationary_distribution(Q), support = States),
+	n = trees[i].states.length
+    )
+    
+    // Propose node_age (new node age) from prior 
+    t ~ Exponential(rate = 20.0)
+    parental_age = max_age + t 
+
+    // Combine picked pair of trees into a new node
+    parental_node = node(left = trees[i], right = trees[j], age = parental_age, states = s)
+
+    // Observe states iid from continuous time Markov chain (CTMC)
+    // Note that trees[i].states can be observed if trees[i] is a leaf, otherwise it
+    // is a latent variable
+    // trees[i].states ~ iid( CTMC( parental_node.states, Q, parental_node.age - trees[i].age ) )
+    // trees[j].states ~ iid( CTMC( parental_node.states, Q, parental_node.age - trees[j].age ) )
+    // (comprehension style)
+    // These could probably be pre-computed
+    somehow because it is only 4 characters,
+    // and always the same time. So this is not the most efficient way to write it.
+    trees[i].states ~ Infer(
+	ctmc(
+	    parental_state = parental_node.states[k],
+	    evolution_time = parental_node.age - trees[i].age,
+	    Q = Q,
+	    P = transition_probabilities(Q)
+	)
+    ) where k in 0:(new_node.states.length - 1)
+    
+    trees[j].states ~ Infer(
+	ctmc(
+	    parental_state = parental_node.states[k],
+	    evolution_time = parental_node.age - trees[j].age,
+	    Q = Q,
+	    P = transition_probabilities(Q)
+	)
+    ) where k in 0:(new_node.states.length - 1)
+    
+    // Compute new_trees vector
+    // new_trees = trees.rm( trees[i], trees[j] ).add( new_node )
+    // ...arr1 converts the array to arguments, think unlist in R
+    // -[i,k] removes i-th and k-th position from the array
+    new_trees = [...trees[-[i,k]], new_node]
+    
+    // cluster new_trees
+    cluster(Q = Q, trees = new_trees, max_age = node_age)
+}
+
+
+// Define model
+// this limitations that models are functions of 0 arguments should not be present
+let model = function () => String
+{
+    pi ~ Dirichlet([1, 1, 1, 1]),
+    //Q = scaled_rate_matrix( pi:pi, er: [1,1], state_labels: ["0","1"] )
+    
+    Q = gtr(p = pi, r = rep(1, 6))
```

Comment:
Typically we would want to infer r, so r would be drawn from a suitable prior. For instance a prior on a simplex space, such as a Dirichlet, if the gtr function scales the rate matrix.